content/js2/blocks/iteration/index.md

```diff
@@ -0,0 +1,38 @@
++++
+title = '🔁 iterating'
+headless = true
+time = 20
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Define iteration'
+    2='Explain why iteration is relevant in a particular problem'
++++
+
+To solve the sub-goal, we have to repeatedly increment the `total` for each number in the array. In programming, the process of repeating something is called **iteration**.
+
+In programming, we can **iterate** by using a {{<tooltip title="loop">}}A loop is a sequence of instructions that is continually repeated until some condition is reached.{{</tooltip>}}.
+
+In particular, we can use a [`for...of` statement](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) to sum the elements of the array.
+
+```js
+function calculateMean(list) {
+  let total = 0;
+  for (const item of list) {
+    total += item;
+  }
+}
+```
+
+{{<tabs name="Playing computer">}}
+
+{{<tab name="🕹️ Play computer">}}
+
+**Play computer** with the code below and explore what happens the `for...of` loop is executed.
+
+Pay close attention to what happens when we enter the `for...of` loop
+
+<iframe title="for..of loop" width="800" height="500" frameborder="0" src="https://pythontutor.com/iframe-embed.html#code=function%20calculateMean%28list%29%20%7B%0A%20%20let%20total%20%3D%200%3B%0A%20%20for%20%28const%20item%20of%20list%29%20%7B%0A%20%20%20%20total%20%2B%3D%20item%3B%0A%20%20%7D%0A%7D%0A%0AcalculateMean%28%5B10,20,30,40,50%5D%29%3B&codeDivHeight=400&codeDivWidth=350&cumulative=false&curInstr=0&heapPrimitives=nevernest&origin=opt-frontend.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false"> </iframe>
```

Comment:
The frames here suggest a distinct `item` variable is declared in the function scope - I don't _think_ this is the case? Do you know what the deal is here?

Reply:
That looks like a bug to me :(

Reply:
Maybe worth switching to the VSCode debugger?

Reply:
Yeah, I followed up, it's a known issue in [the Python Tutor docs](https://docs.google.com/document/d/13_Bc-l2FKMgwPx4dZb0sv7eMfYMHhRVgBRShha8kgbU/edit):
> for-loop variables show up duplicated in two nested blocks due to how the Node.js JavaScript debugger emits its values ([example](https://pythontutor.com/visualize.html#code=//%20%5Bofficial%20Python%20Tutor%20example%5D%0A%0A//%20notice%20how%20there%20are%202%20copies%20of%20the%20i%20and%20j%20loop%20variables%20in%20visualizer%3A%0Afor%20%28let%20i%20%3D%200%3B%20i%20%3C%2010%3B%20i%2B%2B%29%20%7B%0A%20%20for%20%28let%20j%20%3D%200%3B%20j%20%3C%20i%3B%20j%2B%2B%29%20%7B%0A%20%20%20%20console.log%28i,%20j%29%3B%0A%20%20%7D%0A%7D&cumulative=false&heapPrimitives=nevernest&mode=edit&origin=opt-frontend.js&py=js&rawInputLstJSON=%5B%5D&textReferences=false))

If you think there's a convenient way to get folks to use a debugger here, that sounds great, otherwise I guess we can just drop the visualisation (I don't think it's worth including the buggy visualisation and adding a caveat, that'd probably be more confusing than it is useful at this stage).

Reply:
Thanks for following this up - yeah the buggy visualisation will just end up being confusing. I'm going to get rid of it 🧹